### PR TITLE
Update OBPP-03

### PR DESCRIPTION
**Pull request by justusranvier**
- Add BIP number as assigned by BIP editor
- Clarify terms to address comments by BIP editor
- Add an explicit rules section to define the acceptance conditions of alternate form HITs
- Revise abstract

### CODE REVIEW

File: bips/obpp-03.mediawiki [hunk 1]
```diff
@@ -10,20 +10,22 @@
 
 ==Abstract==
 
-The privacy of Bitcoin users with respect to graph analysis is reduced when a transaction is created that contains inputs composed from different scripts. However, creating such transactions is often unavoidable.
+When a Bitcoin transaction contains inputs that reference outputs sent to different Bitcoin addresses, personally identifiable information of user will leak into the blockchain in an uncontrolled manner. While undersirable, these transactions are frequently unavoidable to the natural fragmentation of wallet balances over time.
```
Comment (kristovatlas) [inline bips/obpp-03.mediawiki]: > While **undesirable**, these transactions are frequently unavoidable **due** to the natural fragmentation of wallet balances over time.

Comment (kristovatlas) [inline bips/obpp-03.mediawiki]: Suggested wording:

> When a Bitcoin transaction contains inputs that reference **previous transaction outputs** sent to different Bitcoin addresses

File: bips/obpp-03.mediawiki [hunk 2]
```diff
@@ -10,20 +10,22 @@
...
+This document proposes a set of best practice guidelines which minimize the uncontrolled disclosure of personally identifiable information by defining standard forms for transactions containing heterogenous input scripts.
 
 ==Copyright==
 
 This BIP is in the public domain.
 
 ==Definitions==
 
-* '''Heterogenous input script transaction (HIT)''': A transaction containing multiple inputs where not all inputs have identical scripts (e.g. a transaction spending from more than one Bitcoin address)
+* '''Heterogenous input script transaction (HIT)''': A transaction containing multiple inputs where the scripts of the previous outputs being consumed do not have identical scripts (e.g. a transaction spending outputs which were sent to more than one Bitcoin address)
```
Comment (kristovatlas) [inline bips/obpp-03.mediawiki]: Suggested wording, if still technically correct:

> A transaction containing multiple inputs where the scripts of the previous **transaction** outputs being consumed **are not identical** (e.g. a transaction spending outputs which were sent to more than one Bitcoin address)

File: bips/obpp-03.mediawiki [hunk 3]
```diff
@@ -10,20 +10,22 @@
 
 ==Abstract==
 
-The privacy of Bitcoin users with respect to graph analysis is reduced when a transaction is created that contains inputs composed from different scripts. However, creating such transactions is often unavoidable.
+When a Bitcoin transaction contains inputs that reference previous transaction outputs sent to different Bitcoin addresses, personally identifiable information of user will leak into the blockchain in an uncontrolled manner. While undesirable, these transactions are frequently unavoidable due the natural fragmentation of wallet balances over time.
```
Comment (kristovatlas) [inline bips/obpp-03.mediawiki]: > personally identifiable information of **users**
> or
> personally identifiable information of **the user**

File: bips/obpp-03.mediawiki [hunk 4]
```diff
@@ -71,10 +73,27 @@ When one of the following cases exist, a compliant implementation may create an
 ## Add a spend output to the transaction.
 ## Add a change output to the transaction containing the difference between the current set of inputs and the desired spend.
 # Repeat step 1 to create a second spend output and change output.
+# Repeat step 2 zero or more times.
```
Comment (kristovatlas) [inline bips/obpp-03.mediawiki]: Why?

File: bips/obpp-03.mediawiki [hunk 5]
```diff
@@ -71,10 +73,27 @@ When one of the following cases exist, a compliant implementation may create an
...
 # Adjust the change outputs as necessary to pay the desired transaction fee.
 
 Clients which create intentional HITs must have the capability to form alternate form HITs, and must do so for a non-zero fraction of the transactions they create.
 
+===Rules===
+
+A HIT formed via the proceding procedure will adhere to the following conditions:
```
Comment (kristovatlas) [inline bips/obpp-03.mediawiki]: Suggested wording:

> **An** HIT formed via the **preceding** procedure will adhere to the following conditions:

File: bips/obpp-03.mediawiki [hunk 6]
```diff
@@ -71,10 +73,27 @@ When one of the following cases exist, a compliant implementation may create an
...
+
+# The number of unique inputs scripts must exceed the number of output scripts.
+# All output scripts must be unique.
+# At least one pair of outputs must be of equal value.
+## "Standard outputs" refers to the set of outputs with equal value
+## "Standard value" refers to the value of the standard outputs
+## "Change outputs" refers to all outputs which are not standard outputs
+# For a HIT containing n standard outputs, there must exist at least one possible way to organize the inputs and outputs into n sets, where all sets satisfy the following:
```
Comment (kristovatlas) [inline bips/obpp-03.mediawiki]: I'm a bit confused here. You cannot take n outputs and put them into n disjoint groups that contain at least 2 outputs (spend, change).

File: bips/obpp-03.mediawiki [hunk 7]
```diff
@@ -71,10 +73,27 @@ When one of the following cases exist, a compliant implementation may create an
...
+## The set contains one or more inputs, exactly one standard output, and exactly one change output
+## An input or output that appears in one set must not appear in any other set
+## The sum of the inputs in the set minus the value of the change output is equal to the standard value with a tolerance equal to the transaction fee.
+## Change outputs with a value of zero (virtual change outputs) are permitted. The are defined for the purpose of testing whether or not a HIT adheres to this specification but are not present in the version of the transaction which is broadcast to the network.
+
```
Comment (kristovatlas) [inline bips/obpp-03.mediawiki]: This section could either use some re-wording for clarification, or some examples to illustrate them.

File: bips/obpp-03.mediawiki (hunk 4 above)
Comment (justusranvier) [inline bips/obpp-03.mediawiki]: To allow for intentional HITs to optionally allow more than 2 participants.

File: bips/obpp-03.mediawiki (hunk 6 above)
Comment (justusranvier) [inline bips/obpp-03.mediawiki]: The transaction will have n standard outputs and n change outputs for a total of 2n outputs per HIT.

File: bips/obpp-03.mediawiki (hunk 4 above)
Comment (kristovatlas) [inline bips/obpp-03.mediawiki]: Should this BIP recommend a method of choosing the number of times to repeat this step?

Comment (justusranvier) [inline bips/obpp-03.mediawiki]: "Repeat step 2 zero or more times if desired to increase the number of inputs and/or outputs in the transaction."

**Comment by kristovatlas** [comment]
ACK

**Pull request merged**